cluster/kubernetes/kubernetes.go

```diff
@@ -207,40 +215,39 @@ func (c *Cluster) AllControllers(namespace string) (res []cluster.Controller, er
 // Sync performs the given actions on resources. Operations are
 // asynchronous, but serialised.
 func (c *Cluster) Sync(spec cluster.SyncDef) error {
-	errc := make(chan error)
+	c.mu.Lock()
+	defer c.mu.Unlock()
 	logger := log.With(c.logger, "method", "Sync")
-	c.actionc <- func() {
-		errs := cluster.SyncError{}
-		for _, action := range spec.Actions {
-			logger := log.With(logger, "resource", action.ResourceID)
-			if len(action.Delete) > 0 {
-				obj, err := definitionObj(action.Delete)
-				if err == nil {
-					err = c.applier.Delete(logger, obj)
-				}
-				if err != nil {
-					errs[action.ResourceID] = err
-					continue
-				}
+
+	cs := makeChangeSet()
+	errs := cluster.SyncError{}
+	for _, action := range spec.Actions {
+		stages := []struct {
+			b   []byte
+			cmd string
+		}{
+			{action.Delete, "delete"},
+			{action.Apply, "apply"},
+		}
+		for _, stage := range stages {
+			if len(stage.b) == 0 {
+				continue
 			}
-			if len(action.Apply) > 0 {
-				obj, err := definitionObj(action.Apply)
-				if err == nil {
-					err = c.applier.Apply(logger, obj)
-				}
-				if err != nil {
-					errs[action.ResourceID] = err
-					continue
-				}
+			obj, err := definitionObj(stage.b)
+			id := action.ResourceID
+			if err == nil {
+				cs.stage(stage.cmd, id, obj)
+			} else {
+				errs[id] = err
+				break
 			}
 		}
-		if len(errs) > 0 {
-			errc <- errs
-		} else {
-			errc <- nil
-		}
 	}
-	return <-errc
+	c.applier.apply(logger, cs, errs)
```

Comment:
You don't need the lock until here, I believe

Reply:
Right you are